### PR TITLE
WELD-1838 Revise the use of reflection in Weld SE and Servlet

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

**Comment by weld-tester** [comment]
Triggering build using a merge of e03d57f5122c5382d82ad2ab33f5360abecb096b on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/

**Comment by weld-tester** [comment]
Build 426 is now running using a merge of e03d57f5122c5382d82ad2ab33f5360abecb096b on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/426

**Comment by weld-tester** [comment]
Build 426 outcome was FAILURE using a merge of e03d57f5122c5382d82ad2ab33f5360abecb096b on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/426
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/426

**Comment by weld-tester** [comment]
Triggering build using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/

**Comment by weld-tester** [comment]
Build 427 is now running using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/427

**Comment by weld-tester** [comment]
Build 427 outcome was SUCCESS using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/427
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/427

**Comment by weld-tester** [comment]
Build 427 outcome was SUCCESS using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/427
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/427

**Comment by weld-tester** [comment]
Build 427 outcome was SUCCESS using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/427
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/427

**Comment by weld-tester** [comment]
Build 427 outcome was SUCCESS using a merge of b251310b68bf7a6b3b0753bc2cc02f0dfb500387 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/427
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/427

File: environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/DiscoveryStrategyFactory.java [hunk 1]
```diff
@@ -45,10 +46,15 @@ private DiscoveryStrategyFactory() {
      * @return the discovery strategy
      */
     public static DiscoveryStrategy create(ResourceLoader resourceLoader, Bootstrap bootstrap, Set<Class<? extends Annotation>> initialBeanDefiningAnnotations) {
-
-        if (Reflections.isClassLoadable(JANDEX_INDEX_CLASS_NAME, resourceLoader)) {
+        if (Reflections.isClassLoadable(resourceLoader, JANDEX_INDEX_CLASS_NAME)) {
             CommonLogger.LOG.usingJandex();
-            return Reflections.newInstance(resourceLoader, JANDEX_DISCOVERY_STRATEGY_CLASS_NAME, resourceLoader, bootstrap, initialBeanDefiningAnnotations);
+            try {
+                return Reflections.cast(SecurityActions.getConstructor(Reflections.classForName(resourceLoader, JANDEX_DISCOVERY_STRATEGY_CLASS_NAME),
```
Comment (jharting) [inline environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/DiscoveryStrategyFactory.java]: We should not need to use SecurityActions here as we are obtaining a public constructor of a public class.

Comment (mkouba) [inline environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/DiscoveryStrategyFactory.java]: Good point.

File: environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java [hunk 2]
```diff
@@ -71,14 +67,8 @@ public boolean touch(ContainerContext context) throws Exception {
...
-            Class<?> clazz = Reflections.classForName(JettyWeldInjector.class.getName());
-            Object injector = clazz.getConstructor(WeldManager.class).newInstance(context.getManager());
-            context.getServletContext().setAttribute(INJECTOR_ATTRIBUTE_NAME, injector);
-
-            Class<?> decoratorClass = Reflections.classForName("org.jboss.weld.environment.jetty.WeldDecorator");
-            Method processMethod = decoratorClass.getMethod("process", ServletContext.class);
-            processMethod.invoke(null, context.getServletContext());
-
+            context.getServletContext().setAttribute(INJECTOR_ATTRIBUTE_NAME, new JettyWeldInjector(context.getManager()));
+            WeldDecorator.process(context.getServletContext());
```
Comment (jharting) [inline environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java]: Won't this break in non-Jetty environment?

Comment (mkouba) [inline environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java]: I don't think so. `JettyContainer.initialize(ContainerContext)` will not be called in a non-Jetty environment.

**Comment by weld-tester** [comment]
Triggering build using a merge of 19a93f9a780e412fec0747b4c89787fa286fea6b on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/

**Comment by weld-tester** [comment]
Build 432 outcome was FAILURE using a merge of 19a93f9a780e412fec0747b4c89787fa286fea6b on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/432
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/432

**Comment by mkouba** [comment]
CI is not able to merge the branch. I will do a rebase.

**Comment by weld-tester** [comment]
Triggering build using a merge of 6adc7bd6dd6410873c5f95f1a6af999528a57ab1 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/

**Comment by weld-tester** [comment]
Build 433 is now running using a merge of 6adc7bd6dd6410873c5f95f1a6af999528a57ab1 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/433

**Comment by weld-tester** [comment]
Build 433 outcome was SUCCESS using a merge of 6adc7bd6dd6410873c5f95f1a6af999528a57ab1 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/433
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/433

File: environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java (hunk 2 above)
Comment (jharting) [inline environments/servlet/core/src/main/java/org/jboss/weld/environment/jetty/JettyContainer.java]: OK

**Comment by jharting** [comment]
Looks good. Please send another PR for 2.2

**Comment by weld-tester** [comment]
Build 433 outcome was SUCCESS using a merge of 6adc7bd6dd6410873c5f95f1a6af999528a57ab1 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/433
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/433

**Comment by weld-tester** [comment]
Build 433 outcome was SUCCESS using a merge of 6adc7bd6dd6410873c5f95f1a6af999528a57ab1 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/433
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/433

**Comment by mkouba** [comment]
@jharting Ok, please don't merge it yet. I will add one more commit.

**Comment by weld-tester** [comment]
Triggering build using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/

**Comment by weld-tester** [comment]
Build 441 is now running using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/441

**Comment by weld-tester** [comment]
Build 441 outcome was SUCCESS using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/441
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/441

**Comment by weld-tester** [comment]
Build 441 outcome was SUCCESS using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/441
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/441

**Comment by weld-tester** [comment]
Build 441 outcome was SUCCESS using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/441
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/441

**Comment by weld-tester** [comment]
Build 441 outcome was SUCCESS using a merge of 3ef01d8eec2820ce3eb1167a6b8e586027aa7893 on branch master:
Private: https://jenkins.mw.lab.eng.bos.redhat.com/hudson/job/Weld-2.x-pull-player-executor-osprey/441
Public: https://hudson.jboss.org/hudson/job/Weld-2.x-pull-player-executor-osprey/441

**Comment by jharting** [comment]
Merged, thanks!

**Pull request closed without merging**